File: ClassPal/ClassPal/Services/NeuralDocumentService.swift
import Foundation
import SwiftData

public struct NeuralSearchResultSnippet: Identifiable {
    public let id: String
    public let docId: String
    public let title: String
    public let snippet: String
    public let similarityScore: Double
    public let keywords: [String]

    public init(id: String = UUID().uuidString, docId: String, title: String, snippet: String, similarityScore: Double, keywords: [String] = []) {
        self.id = id
        self.docId = docId
        self.title = title
        self.snippet = snippet
        self.similarityScore = similarityScore
        self.keywords = keywords
    }
}

public final class NeuralDocumentService {
    public static let shared = NeuralDocumentService()

    private let vectorDimension: Int = 64
    private var projectionMatrix: [[Double]] = []
    private var documentIndex: [String: (title: String, content: String, vector: [Double])] = [:]

    private init() {
        generateProjectionMatrix()
    }

    /// Initializes a random projection matrix for 64-D neural text vector embeddings
    private func generateProjectionMatrix() {
        projectionMatrix = []
        var seed: UInt64 = 42
        for _ in 0..<300 {
            var row: [Double] = []
            for _ in 0..<vectorDimension {
                seed = seed &* 6364136223846793005 &+ 1442695040888963407
                let val = (Double(seed % 1000) / 1000.0 - 0.5) * 2.0
                row.append(val)
            }
            projectionMatrix.append(row)
        }
    }

    /// Tokenizes string into clean word tokens
    public func tokenize(_ text: String) -> [String] {
        let clean = text.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { $0.count > 2 }
        return clean
    }

    /// Computes 64-dimensional neural embedding vector for input text
    public func generateEmbedding(_ text: String) -> [Double] {
        let tokens = tokenize(text)
        var vector = Array(repeating: 0.0, count: vectorDimension)

        guard !tokens.isEmpty else { return vector }

        for token in tokens {
            var hash: Int = 0
            for char in token.utf8 {
                hash = (hash &<< 5) &- hash &+ Int(char)
            }
            let idx = abs(hash) % projectionMatrix.count
            let row = projectionMatrix[idx]
            for d in 0..<vectorDimension {
                vector[d] += row[d]
            }
        }

        // L2 Normalization
        let mag = sqrt(vector.reduce(0.0) { $0 + $1 * $1 })
        if mag > 0 {
            for d in 0..<vectorDimension {
                vector[d] /= mag
            }
        }
        return vector
    }

    /// Cosine similarity calculation
    public func cosineSimilarity(_ vecA: [Double], _ vecB: [Double]) -> Double {
        guard vecA.count == vecB.count, !vecA.isEmpty else { return 0.0 }
        var dot = 0.0
        var normA = 0.0
        var normB = 0.0
        for i in 0..<vecA.count {
            dot += vecA[i] * vecB[i]
            normA += vecA[i] * vecA[i]
            normB += vecB[i] * vecB[i]
        }
        if normA == 0 || normB == 0 { return 0.0 }
        return dot / (sqrt(normA) * sqrt(normB))
    }

    /// Trains local neural network index on a course document
    public func trainOnDocument(docId: String, title: String, content: String) {
        let vector = generateEmbedding(content)
        documentIndex[docId] = (title: title, content: content, vector: vector)
    }

    /// Performs semantic vector search over learned document index
    public func search(query: String, topK: Int = 5) -> [NeuralSearchResultSnippet] {
        let queryVector = generateEmbedding(query)
        var results: [NeuralSearchResultSnippet] = []

        for (docId, data) in documentIndex {
            let score = cosineSimilarity(queryVector, data.vector)
            let tokens = Array(Set(tokenize(data.content))).prefix(5)
            let snippet = String(data.content.prefix(200))
            results.append(NeuralSearchResultSnippet(
                docId: docId,
                title: data.title,
                snippet: snippet,
                similarityScore: score,
                keywords: Array(tokens)
            ))
        }
        results.sort { $0.similarityScore > $1.similarityScore }
        return Array(results.prefix(topK))
    }

    /// Total indexed document count
    public var indexedDocumentCount: Int {
        return documentIndex.count
    }
}

File: src/services/NeuralDocumentService.ts
/**
 * NeuralDocumentService
 * On-device 64-dimensional semantic text search and document indexer.
 * 1:1 match with Swift NeuralDocumentService
 */

import { generateNeuralProjectionMatrix, cosineSimilarity } from '../utils/mathPhysics';
import { NeuralProjectionConstants } from '../constants/physics';

export interface NeuralSearchResultSnippet {
  id: string;
  docId: string;
  title: string;
  snippet: string;
  similarityScore: number;
  keywords: string[];
}

export interface IndexedDocument {
  title: string;
  content: string;
  vector: number[];
}

export class NeuralDocumentService {
  private static _instance: NeuralDocumentService;
  public static get shared(): NeuralDocumentService {
    if (!this._instance) {
      this._instance = new NeuralDocumentService();
    }
    return this._instance;
  }

  private readonly vectorDimension: number = NeuralProjectionConstants.vectorDimension;
  private projectionMatrix: number[][];
  private documentIndex: Map<string, IndexedDocument> = new Map();

  public constructor() {
    this.projectionMatrix = generateNeuralProjectionMatrix();
  }

  /**
   * Tokenizes text into clean lowercase alphanumeric words (length > 2)
   */
  public tokenize(text: string): string[] {
    return text
      .toLowerCase()
      .split(/[^a-z0-9]+/)
      .filter(token => token.length > 2);
  }

  /**
   * Computes 64-dimensional neural embedding vector for input text
   */
  public generateEmbedding(text: string): number[] {
    const tokens = this.tokenize(text);
    const vector = new Array(this.vectorDimension).fill(0.0);

    if (tokens.length === 0) {
      return vector;
    }

    const rowCount = BigInt(this.projectionMatrix.length);

    for (const token of tokens) {
      let hash = 0n;
      for (let i = 0; i < token.length; i++) {
        const code = BigInt(token.charCodeAt(i));
        hash = ((hash << 5n) - hash + code) & 0x7FFFFFFFFFFFFFFFn;
      }
      const idx = Number(hash % rowCount);
      const row = this.projectionMatrix[idx];
      for (let d = 0; d < this.vectorDimension; d++) {
        vector[d] += row[d];
      }
    }

    // L2 Normalization
    let sumSq = 0.0;
    for (let d = 0; d < this.vectorDimension; d++) {
      sumSq += vector[d] * vector[d];
    }
    const mag = Math.sqrt(sumSq);
    if (mag > 0) {
      for (let d = 0; d < this.vectorDimension; d++) {
        vector[d] /= mag;
      }
    }

    return vector;
  }

  /**
   * Calculates cosine similarity between two 64-D vectors
   */
  public cosineSimilarity(vecA: number[], vecB: number[]): number {
    return cosineSimilarity(vecA, vecB);
  }

  /**
   * Trains/indexes a document in memory
   */
  public trainOnDocument(docId: string, title: string, content: string): void {
    const vector = this.generateEmbedding(content);
    this.documentIndex.set(docId, { title, content, vector });
  }

  /**
   * Semantic vector search across indexed documents
   */
  public search(query: string, topK: number = 5): NeuralSearchResultSnippet[] {
    const queryVector = this.generateEmbedding(query);
    const results: NeuralSearchResultSnippet[] = [];

    for (const [docId, data] of this.documentIndex.entries()) {
      const score = this.cosineSimilarity(queryVector, data.vector);
      const uniqueTokens = Array.from(new Set(this.tokenize(data.content))).slice(0, 5);
      const snippet = data.content.substring(0, 200);

      results.push({
        id: `search-${Math.random().toString(36).substring(2, 10)}`,
        docId,
        title: data.title,
        snippet,
        similarityScore: score,
        keywords: uniqueTokens
      });
    }

    results.sort((a, b) => b.similarityScore - a.similarityScore);
    return results.slice(0, topK);
  }

  public clearIndex(): void {
    this.documentIndex.clear();
  }

  public get indexedDocumentCount(): number {
    return this.documentIndex.size;
  }
}
